google/cloud/security/common/gcp_api/_base_repository.py

```diff
@@ -0,0 +1,695 @@
+        return method(prior_request, prior_response)
+
+    def _request_supports_pagination(self, verb):
+        """Determines if the API action supports pagination.
+
+        Args:
+          verb (str): Request verb (ex. insert, update, delete).
+
+        Returns:
+          bool: True when API supports pagination, False otherwise.
+        """
+        return getattr(self._component, verb + '_next', None)
+
+    def execute_command(self, verb, verb_arguments):
+        """Executes command (ex. add) via a dedicated http object.
+
+        Async APIs may take minutes to complete. Therefore, callers are
+        encouraged to leverage concurrent.futures (or similar) to place long
+        running commands on a separate threads.
+
+        Args:
+          verb (str): Method to execute on the component (ex. get, list).
+          verb_arguments (dict): key-value pairs to be passed to
+              _build_request.
+
+        Returns:
+          dict: An async operation Service Response.
+        """
+        request = self._build_request(verb, verb_arguments)
+        request_submission_status = self._execute(request)
+
+        return request_submission_status
+
+    def execute_paged_query(self, verb, verb_arguments):
+        """Executes query (ex. list) via a dedicated http object.
+
+        Args:
+          verb (str): Method to execute on the component (ex. get, list).
+          verb_arguments (dict): key-value pairs to be passed to _BuildRequest.
+
+        Yields:
+          dict: Service Response.
+
+        Raises:
+          PaginationNotSupportedError: When an API does not support paging.
+        """
+        if not self._request_supports_pagination(verb=verb):
+            raise api_errors.PaginationNotSupportedError(
+                '{} does not support pagination')
+
+        request = self._build_request(verb, verb_arguments)
+
+        number_of_pages_processed = 0
+        while request is not None:
+            response = self._execute(request)
+            number_of_pages_processed += 1
+            LOGGER.debug('Executing paged request #%s',
+                         number_of_pages_processed)
+            request = self._build_next_request(verb, request, response)
+            yield response
+
+    def execute_search_query(self, verb, verb_arguments):
+        """Executes query (ex. search) via a dedicated http object.
+
+        Args:
+          verb (str): Method to execute on the component (ex. search).
+          verb_arguments (dict): key-value pairs to be passed to _BuildRequest.
+
+        Yields:
+          dict: Service Response.
+        """
+        # Implementation of search does not follow the standard API pattern.
+        # Fields need to be in the body rather than sent seperately.
+        next_page_token = None
+        number_of_pages_processed = 0
+        while True:
+            req_body = verb_arguments.get('body', dict())
+            if next_page_token:
```

Comment:
This looks very similar to paged_query, why do we need both?

Reply:
The search APIs require nextPageToken to be inside of a body argument, not a top level parameter. And they do not implement the {verb}_next() method. So they need a custom implementation.